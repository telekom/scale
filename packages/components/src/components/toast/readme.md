s# t-toast

<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description                          | Type                | Default     |
| --------------- | ---------------- | ------------------------------------ | ------------------- | ----------- |
| `animated`      | `animated`       | (optional) Animated toast            | `boolean`           | `true`      |
| `autoHide`      | `auto-hide`      | (optional) Toast autohide time       | `boolean \| number` | `false`     |
| `customClass`   | `custom-class`   | (optional) Toast class               | `string`            | `''`        |
| `fadeDuration`  | `fade-duration`  | (optional) Toast fade duration       | `number`            | `500`       |
| `opened`        | `opened`         | (optional) Toast opened              | `boolean`           | `undefined` |
| `positionRight` | `position-right` | (optional) Toast position right      | `number`            | `12`        |
| `positionTop`   | `position-top`   | (optional) Toast position at the top | `number`            | `12`        |
| `size`          | `size`           | (optional) Toast size                | `string`            | `''`        |
| `theme`         | `theme`          | (optional) Toast theme               | `string`            | `''`        |
| `time`          | `time`           | (optional) Toast time                | `number`            | `undefined` |
| `variant`       | `variant`        | (optional) Toast variant             | `string`            | `''`        |


## Methods

### `open() => Promise<void>`

Toast method: open()

#### Returns

Type: `Promise<void>`




## CSS Custom Properties

| Name                               | Description                       |
| ---------------------------------- | --------------------------------- |
| `--toast-background`               | Background color of the toast     |
| `--toast-body-margin-bottom`       | Margin bottom of the toast body   |
| `--toast-body-padding`             | Padding of the toast body         |
| `--toast-border-radius`            | Border radius of the toast        |
| `--toast-box-shadow`               | Box shadow of the toast           |
| `--toast-header-border-bottom`     | Border bottom of the toast header |
| `--toast-header-font-size`         | Font size of the toast header     |
| `--toast-header-margin`            | Margin of the toast header        |
| `--toast-header-padding`           | Padding of the toast header       |
| `--toast-header-small-font-size`   | Font size of the small header     |
| `--toast-header-small-margin-left` | Margin left of the small header   |
| `--toast-header-small-margin-top`  | Margin top of the small header    |
| `--toast-height`                   | Height of the toast               |
| `--toast-width`                    | Width of the toast                |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
