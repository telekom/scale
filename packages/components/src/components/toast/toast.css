/**
* @prop --toast-width: Width of the toast
* @prop --toast-height: Height of the toast
* @prop --toast-border-radius: Border radius of the toast
* @prop --toast-box-shadow: Box shadow of the toast
* @prop --toast-background: Background color of the toast
* @prop --toast-body-padding: Padding of the toast body
* @prop --toast-body-margin-bottom: Margin bottom of the toast body
* @prop --toast-header-padding: Padding of the toast header
* @prop --toast-header-border-bottom: Border bottom of the toast header
* @prop --toast-header-font-size: Font size of the toast header
* @prop --toast-header-margin: Margin of the toast header
* @prop --toast-header-small-font-size: Font size of the small header
* @prop --toast-header-small-margin-left: Margin left of the small header
* @prop --toast-header-small-margin-top: Margin top of the small header
*/

.toast {
  box-sizing: border-box;
  width: var(--toast-width, 400px);
  position: fixed;
  border-radius: var(--toast-border-radius, 3px);
  box-shadow: var(--toast-box-shadow, 0 2px 8px rgba(0, 0, 0, 0.26));
  background: var(--toast-background, #fff);
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  z-index: 1;
}

.toast__body {
  padding: var(--toast-body-padding, 1rem);
}

.toast__header {
  padding: var(--toast-header-padding, 1rem);
  border-bottom: var(--toast-header-border-bottom, 1px solid #ccc);
  font-size: var(--toast-header-font-size, 1.25rem);
  margin: var(--toast-header-margin, 0);
  justify-content: space-between;
  display: flex;
}

.toast__header a {
  cursor: pointer;
}

.toast__header small {
  font-size: var(--toast-header-small-font-size, 0.8rem);
  margin-left: var(--toast-header-small-margin-left, 120px);
  margin-top: var(--toast-header-small-margin-top, 5px);
}

.toast__progress {
  background: red;
  height: 2px;
  display: block;
  position: absolute;
  bottom: 0;
  left: 0;
  overflow: hidden;
}
