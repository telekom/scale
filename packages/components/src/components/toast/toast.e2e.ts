import { newE2EPage } from '@stencil/core/testing';

describe('t-toast', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<t-toast>Toast message</t-toast>');
    const element = await page.find('t-toast');
    expect(element).toHaveClass('hydrated');
  });
});
