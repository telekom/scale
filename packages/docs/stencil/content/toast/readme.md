# scale-toast



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute        | Description                          | Type                                     | Default     |
| --------------- | ---------------- | ------------------------------------ | ---------------------------------------- | ----------- |
| `animated`      | `animated`       | (optional) Animated toast            | `boolean`                                | `true`      |
| `autoHide`      | `auto-hide`      | (optional) Toast autohide time       | `boolean \| number`                      | `false`     |
| `customClass`   | `custom-class`   | (optional) Toast class               | `string`                                 | `''`        |
| `fadeDuration`  | `fade-duration`  | (optional) Toast fade duration       | `number`                                 | `500`       |
| `opened`        | `opened`         | (optional) Toast opened              | `boolean`                                | `undefined` |
| `positionRight` | `position-right` | (optional) Toast position right      | `number`                                 | `12`        |
| `positionTop`   | `position-top`   | (optional) Toast position at the top | `number`                                 | `12`        |
| `size`          | `size`           | (optional) Toast size                | `string`                                 | `''`        |
| `styles`        | --               | (optional) Injected jss styles       | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `time`          | `time`           | (optional) Toast time                | `number`                                 | `undefined` |
| `variant`       | `variant`        | (optional) Toast variant             | `string`                                 | `''`        |


## Methods

### `open() => Promise<void>`

Toast method: open()

#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
