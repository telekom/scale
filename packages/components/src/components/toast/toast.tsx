import {
  Component,
  Prop,
  Method,
  h,
  State,
  Element,
  Host,
} from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { formatDistance, subSeconds } from 'date-fns';
import { styles } from './toast.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

@Component({
  tag: 'scale-toast',
  shadow: true,
})
export class Toast implements Base {
  /** (optional) Toast class */
  @Prop() customClass?: string = '';
  /** (optional) Toast size */
  @Prop() size?: string = '';
  /** (optional) Toast variant */
  @Prop() variant?: string = '';
  /** (optional) Toast opened */
  @Prop({ reflectToAttr: true }) opened?: boolean;
  /** (optional) Toast autohide time */
  @Prop() autoHide?: boolean | number = false;
  /** (optional) Animated toast */
  @Prop() animated?: boolean = true;
  /** (optional) Toast time */
  @Prop() time?: number;
  /** (optional) Toast position at the top */
  @Prop() positionTop?: number = 12;
  /** (optional) Toast position right */
  @Prop() positionRight?: number = 12;
  /** (optional) Toast fade duration */
  @Prop() fadeDuration?: number = 500;

  /** (optional) Toast state progress */
  @State() progress: number = 0;
  /** (optional) Toast state height with offset */
  @State() toastHeightWithOffset: number = 0;

  @Element() element: HTMLElement;

  /** (optional) Injected jss styles */
  @Prop({ reflectToAttr: true }) styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Toast', styles) stylesheet: StyleSheet;

  hideToast: boolean = false;

  timerId = null;

  componentDidUnload() {
    if (this.timerId) {
      clearTimeout(this.timerId);
      this.timerId = null;
      this.opened = false;
      this.progress = 0;
    }
  }
  componentWillUpdate() {}

  close = () => {
    clearInterval(this.timerId);
    this.hideToast = true;
    setTimeout(() => {
      this.timerId = null;
      this.opened = false;
      this.progress = 0;
    }, this.fadeDuration);
  };

  getTime = () => {
    const formattedTime =
      this.time &&
      formatDistance(subSeconds(this.time, 3), new Date(), { addSuffix: true });
    return formattedTime;
  };

  setToastTimeout = () => {
    if (this.opened && this.autoHide !== false && !this.timerId) {
      this.timerId = setInterval(() => {
        this.progress += 1 / (this.getAutoHide() / 1000);
        if (this.progress >= 100) {
          this.close();
        }
      }, 10);
    }
  };

  /** Toast method: open() */
  @Method()
  async open() {
    this.opened = true;
    this.hideToast = false;
  }

  render() {
    const { classes } = this.stylesheet;
    this.setToastTimeout();
    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <style>{this.transitions(this.toastHeightWithOffset)}</style>
        <style>{this.animationStyle(this.toastHeightWithOffset)}</style>
        <div class={this.getCssClassMap()}>
          <div class={classes.toast__header}>
            <slot name="header" />
            header
            <small>{this.getTime()}</small>
            <a onClick={this.close}>
              <span aria-hidden="true">&times;</span>
            </a>
          </div>
          {this.autoHide && (
            <div
              class={classes.toast__progress}
              style={{ width: `${this.progress}%` }}
            >
              &nbsp;
            </div>
          )}
          <div class={classes.toast__body}>
            <slot />
          </div>
        </div>
      </Host>
    );
  }

  transitions = offset => `
    @keyframes fadeIn {
      from {
        opacity: 0;
        top: -${offset}px;
      }
      to {
        opacity: 1;
        top: ${this.positionTop}px;
      }
    }

    @keyframes fadeOut {
      from {
        opacity: 1;
        top: ${this.positionTop}px;
      }
      to {
        opacity: 0;
        top: -${offset}px;
      }
    }
  `;

  animationStyle = offset => {
    this.stylesheet.addRule('toast--show', {
      right: `${this.positionRight}px`,
      animation: `fadeIn ${this.fadeDuration / 1000}s ease-in-out`,
      top: `${this.positionTop}px`,
      opacity: 1,
    });
    this.stylesheet.addRule('toast--hide', {
      right: `${this.positionRight}px`,
      animation: `fadeOut ${this.fadeDuration / 1000}s ease-in-out`,
      top: `-${offset}px`,
      opacity: 0,
    });
  };

  getToastHeightWithOffset() {
    const { classes } = this.stylesheet;
    const toastHeight = this.element.shadowRoot.querySelector(
      `.${classes.toast}`
    ).scrollHeight;
    this.toastHeightWithOffset = toastHeight + this.positionTop;
  }

  getAutoHide() {
    if (
      typeof this.autoHide === 'number' ||
      typeof this.autoHide === 'string'
    ) {
      return Number(this.autoHide);
    } else {
      return 0;
    }
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.toast,
      this.customClass && this.customClass,
      this.size && classes[`toast--size-${this.size}`],
      this.variant && classes[`toast--variant-${this.variant}`],
      !!this.opened && classes[`toast--opened`],
      !!!this.hideToast && classes[`toast--show`],
      !!this.hideToast && classes[`toast--hide`]
    );
  }
}
