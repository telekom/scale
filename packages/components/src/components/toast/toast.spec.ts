import { newSpecPage } from '@stencil/core/testing';
import { Toast } from './toast';

describe('Toast', () => {
  let element;
  beforeEach(async () => {
    element = new Toast();
    jest.useFakeTimers();
    jest.mock('date-fns');
  });

  const components = [Toast];

  const timeStamp = 1540035262000;

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components,
      html: `<t-toast>Toast message</t-toast>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot when opened', async () => {
    const page = await newSpecPage({
      components,
      html: `<t-toast opened=true >Label</t-toast>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should close the Toast', () => {
    expect(element.opened).toBe(undefined);
    element.close();
    setTimeout(() => {
      expect(element.opened).toBe(false);
    }, 10);
  });

  it('should open the Toast', () => {
    expect(element.opened).toBe(undefined);
    element.open();
    expect(element.opened).toBe(true);
  });

  it('should hide the toast', () => {
    element.autohide = true;
    element.opened = true;
    element.setToastTimeout();

    setTimeout(() => {
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(setTimeout).toHaveBeenLastCalledWith(
        expect.any(Function),
        element.autohideTime
      );
    }, 10);
  });

  it('should not hide the toast', () => {
    element.autohide = false;
    element.opened = false;
    element.setToastTimeout();
    expect(element.autohide).toBe(false);
  });

  it('should cancel the timeout', () => {
    element.myTimeout = 500;
    element.close();

    setTimeout(() => {
      expect(clearTimeout).toHaveBeenCalledTimes(1);
      expect(element.myTimeout).toEqual(undefined);
    }, 10);
  });

  it('should have a default css class', () => {
    expect(element.getCssClassMap()).toBe('toast');
  });

  it('should handle custom css class', () => {
    element.customClass = 'custom-class';
    expect(element.getCssClassMap()).toContain('custom-class');
  });

  it('should handle size css class', () => {
    element.size = 'small';
    expect(element.getCssClassMap()).toContain('toast--size-small');
  });

  it('should handle theme css class', () => {
    element.theme = 'default';
    expect(element.getCssClassMap()).toContain('toast--theme-default');
  });

  it('should handle variant css class', () => {
    element.variant = 'primary';
    expect(element.getCssClassMap()).toContain('toast--variant-primary');
  });

  it('should render with default timeformat', () => {
    element.time = timeStamp;
    element.getTime();

    expect(element.time).toBe(timeStamp);
  });
});
