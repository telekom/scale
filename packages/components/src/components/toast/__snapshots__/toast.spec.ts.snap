// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should match snapshot 1`] = `
<t-toast>
  <mock:shadow-root>
    <div class="toast">
      <style>
        .toast {
      right: 12px;
      top: -NaNpx;
    }

    .toast--show {
      animation: fadeIn 0.5s ease-in-out;
      animation-timing-function: ease-out;
      top: 12px;
    }

    .toast--hide {
      animation: fadeOut 0.5s ease-in-out;
      animation-timing-function: ease-in;
    }

    @keyframes fadeIn {
      from {
        opacity: 0;
        top: -NaNpx;
      }
      to {
        opacity: 1;
        top: 12px;
      }
    }

    @keyframes fadeOut {
      from {
        opacity: 1;
        top: 12px;
      }
      to {
        opacity: 0;
        top: -NaNpx;
      }
    }
      </style>
      <div class="toast__header">
        <slot name="header"></slot>
        header
        <small></small>
        <a>
          <span aria-hidden="true">
            ×
          </span>
        </a>
      </div>
      <div class="toast__body">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  Toast message
</t-toast>
`;

exports[`Toast should match snapshot when opened 1`] = `
<t-toast opened="">
  <mock:shadow-root>
    <div class="toast toast--show">
      <style>
        .toast {
      right: 12px;
      top: -NaNpx;
    }

    .toast--show {
      animation: fadeIn 0.5s ease-in-out;
      animation-timing-function: ease-out;
      top: 12px;
    }

    .toast--hide {
      animation: fadeOut 0.5s ease-in-out;
      animation-timing-function: ease-in;
    }

    @keyframes fadeIn {
      from {
        opacity: 0;
        top: -NaNpx;
      }
      to {
        opacity: 1;
        top: 12px;
      }
    }

    @keyframes fadeOut {
      from {
        opacity: 1;
        top: 12px;
      }
      to {
        opacity: 0;
        top: -NaNpx;
      }
    }
      </style>
      <div class="toast__header">
        <slot name="header"></slot>
        header
        <small></small>
        <a>
          <span aria-hidden="true">
            ×
          </span>
        </a>
      </div>
      <div class="toast__body">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  Label
</t-toast>
`;
